### PR TITLE
ENt-3992 | added check for linked identity provider.

### PR DESCRIPTION
**Pull request by HammadAhmadWaqas**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/models.py [hunk 1]
```diff
@@ -366,7 +366,7 @@ def enterprise_customer_identity_provider(self):  # pylint: disable=invalid-name
     @property
     def identity_provider(self):
         """
-        Return the identity providers associated with this enterprise customer.
+        Return the identity provider id associated with this enterprise customer.
```
Comment (zamanafzal) [inline enterprise/models.py:369]: shouldn't it be `Return the identity **providers** id associated with this enterprise customer.`

File: enterprise/decorators.py [hunk 2]
```diff
@@ -129,10 +129,13 @@ def wrapper(request, *args, **kwargs):
         enterprise_customer = get_enterprise_customer_or_404(enterprise_uuid)
         query_params = request.GET
         # Check if tpa_hint was passed as query param
-        tpa_hint = query_params.get('tpa_hint')
-        if not tpa_hint and not enterprise_customer.has_multiple_idps:
-            tpa_hint = enterprise_customer.identity_provider\
-                if enterprise_customer.identity_provider else None
+        tpa_hint_param = query_params.get('tpa_hint')
```
Comment (zamanafzal) [inline enterprise/decorators.py:132]: Can we please add a comment here. So that it can be understandable why we are doing it.

**Review by zamanafzal — APPROVED** [review]
LGTM.

File: enterprise/models.py (hunk 1 above)
Comment (HammadAhmadWaqas) [reply enterprise/models.py:369]: its returning id of only *one* identity_provider(i.e first one) so think it should 
`Return the identity provider id associated with this enterprise customer.`
or
`Return the identity provider's id associated with this enterprise customer.`

File: enterprise/decorators.py (hunk 2 above)
Comment (HammadAhmadWaqas) [reply enterprise/decorators.py:132]: is not it self-explanatory?

Comment (mattdrayer) [reply enterprise/decorators.py:132]: I think a comment that provides context would be helpful to future collaborators.

File: enterprise/models.py (hunk 1 above)
Comment (mattdrayer) [reply enterprise/models.py:369]: Is this where the check for a "default" IdP would happen?

Comment (mattdrayer) [reply enterprise/models.py:369]: Also this seems like a good place to add some logging information -- for example what if no IdP is found.

File: enterprise/views.py [hunk 3]
```diff
@@ -1071,8 +1071,14 @@ def get(self, request):
 
         # Check if tpa_hint was passed as query param then redirect to that IDP login
         # Otherwise check if enterprise is linked to one IDP, then page will be redirected there
-        tpa_hint = query_params.get('tpa_hint') or enterprise_customer.identity_providers.first().provider_id if \
-            enterprise_customer.identity_providers.count() == 1 else None
+
+        tpa_hint_param = query_params.get('tpa_hint')
```
Comment (mattdrayer) [inline enterprise/views.py:1075]: duplicate code?

File: tests/test_enterprise/views/test_enterprise_proxy_login_view.py [hunk 4]
```diff
@@ -99,28 +99,50 @@ def test_tpa_redirect(self, use_next):
...
+        ('fake_idp', 'idp-1', True, None),
+        ('idp-1', 'idp-1', False, None),
+        (None, None, False, None),
     )
     @ddt.unpack
-    def test_tpa_hint_redirect(self, use_tpa_hint, tpa_hint):
+    def test_tpa_redirects_using_tpa_hint_param(
+            self,
+            tpa_hint_param,
+            identity_provided_id,
```
Comment (mattdrayer) [inline tests/test_enterprise/views/test_enterprise_proxy_login_view.py:117]: should this be `identity_provider_id`?

File: enterprise/models.py [hunk 5]
```diff
@@ -383,6 +383,13 @@ def identity_providers(self):
         """
         return self.enterprise_customer_identity_providers.all()
 
+    @property
+    def identity_provider_ids(self):
+        """
+        Return the identity provider Ids associated with this enterprise customer.
+        """
+        return str(self.identity_providers.values_list('provider_id', flat=True))
```
Comment (irfanuddinahmad) [inline enterprise/models.py:391]: Shouldn't we return a list rather than a string?

Comment (HammadAhmadWaqas) [reply enterprise/models.py:391]: @irfanuddinahmad good catch

File: enterprise/decorators.py (hunk 2 above)
Comment (HammadAhmadWaqas) [reply enterprise/decorators.py:131]: moved this logic to a class method with comments.

File: enterprise/views.py (hunk 3 above)
Comment (HammadAhmadWaqas) [reply enterprise/views.py:1072]: moved into a common method

File: enterprise/models.py (hunk 1 above)
Comment (HammadAhmadWaqas) [reply enterprise/models.py]: Yes, it seems a good place for a default check. added logger.

File: tests/test_enterprise/views/test_enterprise_proxy_login_view.py (hunk 4 above)
Comment (HammadAhmadWaqas) [reply tests/test_enterprise/views/test_enterprise_proxy_login_view.py]: good catch, updated.

**Review by mattdrayer — APPROVED** [review]

**Pull request merged**